---
title: "ABX R21 Preliminary Analysis"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: false
    toc_depth: 3
  BiocStyle::pdf_document:
    toc: true
    toc_float: false
    toc_depth: 3
abstract: "SIAMCAT: Vincent + Raymond (This vignette is part of Chloe's thesis Aim3.)"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, collapse = TRUE, 
                      warning = FALSE, eval = FALSE)
```

# Intro
**Develop and evaluate the prediction models for recent antibiotics usage**    
Researchers often exclude participants with recent antibiotic use, but this may introduce selection bias by excluding the least healthy individuals. *We hypothesize that antibiotic usage can be detected from the current microbiome status using the prediction models trained on microbiome datasets with recent antibiotics exposure.* We will construct the prediction models and test how well they can infer probable antibiotic exposure from the gut microbiome. Using prediction models from this aim, researchers can stratify analyses or include participants who may have otherwise been excluded from the study.

## Load packages
SIAMCAT is used for this analysis.
```{r eval=TRUE}
suppressPackageStartupMessages({
    library(curatedMetagenomicData)
    library(dplyr)
    library(mia)
    library(SIAMCAT)
    library(stringr)
    library(ggplot2)
    library(microbiomeABXR21) # temporary package for EDA
})
```

## SIAMCAT script
This script requires `allData`,`studyName` and `abx` inputs.
```{r}
siamcat_cmd <- system.file("scripts/siamcat_cmd.R", 
                           package = "microbiomeABXR21")
```

## Load datasets
```{r eval=TRUE}
studyNames <- c("VincentC_2016", "RaymondF_2016")

## Metadata
meta <- curatedMetagenomicData::sampleMetadata
submeta <- meta %>% 
    filter(study_name == studyNames[1] | study_name == studyNames[2])
```      

Chloe found out that there is an error in the Vincent metadata in cMD package.
This is the corrected metadata:
```{r}
dir <- "~/Packages/microbiomeABXR21/inst/extdata/"
meta_vi <- read.csv(file.path(dir, "Vincent_metadata.csv"))
```


Prepare `allData` object
```{r allData, eval=TRUE, message=FALSE}
## Download data
allData <- returnSamples(submeta, "relative_abundance")
colData(allData)$study_name %>% table() # summarize the number of samples
assay(allData) <- assay(allData)/100 # percentage to decimals

## Clean ABX info
colData(allData)$abx <- colData(allData)$antibiotics_family
colData(allData)$abx[which(is.na(colData(allData)$abx))] <- "control"
abx_ls <- c("carbapenems", "cephalosporins", "fluoroquinolones",
            "glycopeptides", "macrolides", "nitroimidazoles",
            "penicillin", "sulfonamides")
for (abx_name in abx_ls) {
    ind <- which(colData(allData)$antibiotics_family %>% str_detect(., abx_name))
    colData(allData)$abx[ind] <- abx_name
}
```

```{r ABX_summary, eval=TRUE}
## Check the ABX classes used
for (studyName in studyNames) {
    sub_ind <- colData(allData)$study_name == studyName
    res <- table(colData(allData)$abx[sub_ind])
    print(paste0("ABX classes in ", studyName, ":"))
    print(res)
}
```

# Modeling
## Raymond dataset
```{r message=FALSE, warning=FALSE}
## Required inputs
studyName <- "RaymondF_2016"
abx <- "cephalosporins"
source(siamcat_cmd)
```

```{r echo=FALSE, message=FALSE, eval=TRUE}
studyName <- "RaymondF_2016"
abx <- "cephalosporins"

fname <- paste0(studyName, "_evaluation_", abx)
if (length(metadata(allData)) != 0) {
    fname <- paste0(fname, "_", metadata(allData))}

model_eval <- readRDS(paste0(fname, ".rds"))
model.evaluation.plot(model_eval)
```

## Vincent dataset
```{r message=FALSE, warning=FALSE}
## Required inputs
studyName <- "VincentC_2016"
abx <- "penicillin"
source(siamcat_cmd)

## Error from `check.confounders` step:      
# Error in ctrfn(levels(x), contrasts = contrasts) :
# orthogonal polynomials cannot be represented accurately enough for 97 degrees of freedom
```

```{r echo=FALSE, message=FALSE, eval=TRUE}
studyName <- "VincentC_2016"
abx <- "penicillin"

fname <- paste0(studyName, "_evaluation_", abx)
if (length(metadata(allData)) != 0) {
    fname <- paste0(fname, "_", metadata(allData))}

model_eval <- readRDS(paste0(fname, ".rds"))
model.evaluation.plot(model_eval)
```


# Genus-Level Modeling
```{r eval=TRUE}
## Genus-level data
allData <- agglomerateByRank(allData, rank = "genus")
```

## Raymond dataset
```{r}
studyName <- "RaymondF_2016"
abx <- "cephalosporins"
source(siamcat_cmd)
```

```{r echo=FALSE, message=FALSE, eval=TRUE}
studyName <- "RaymondF_2016"
abx <- "cephalosporins"

fname <- paste0(studyName, "_evaluation_", abx)
if (length(metadata(allData)) != 0) {
    fname <- paste0(fname, "_", metadata(allData))}

model_eval <- readRDS(paste0(fname, ".rds"))
model.evaluation.plot(model_eval)
```

## Vincent dataset
```{r}
studyName <- "VincentC_2016"
abx <- "penicillin"
source(siamcat_cmd)
```

```{r echo=FALSE, message=FALSE, eval=TRUE}
studyName <- "VincentC_2016"
abx <- "penicillin"

fname <- paste0(studyName, "_evaluation_", abx)
if (length(metadata(allData)) != 0) {
    fname <- paste0(fname, "_", metadata(allData))}

model_eval <- readRDS(paste0(fname, ".rds"))
model.evaluation.plot(model_eval)
```
